### PR TITLE
Introduction of oakpal-maven-plugin and oakpal-checks.

### PR DESCRIPTION
**Pull request by adamcin**
This PR adds `acs-internal` and `acs-commons-integrators` [OakPAL checklists](http://adamcin.net/oakpal/oakpal-core/publishing-a-checklist.html) to the acs-aem-commons project. The `acs-internal` checklist is intended to be used to enforce rules for the acs-aem-commons-content package, whereas the `acs-commons-integrators` checklist can be used as a standard checklist for the community of ACS AEM Commons integrators to ensure compatibility with the acs-aem-commons-content package.

The first implemented check is to detect content classification violations described here: https://helpx.adobe.com/experience-manager/6-4/sites/deploying/using/sustainable-upgrades.html#ContentClassifications

The basic checklist is also enabled, resulting in the following report for the acs-aem-commons-content package:

```
[INFO] --- oakpal-maven-plugin:1.1.3:scan (oakpal-scan-aem64) @ acs-aem-commons-content ---
[INFO] Found a new index node [reference]. Reindexing is requested
[INFO] Reindexing will be performed for following indexes: [/oak:index/uuid, /oak:index/reference, /oak:index/nodetype]
[INFO] Indexing report
    - /oak:index/uuid*(0)
    - /oak:index/reference*(0)
    - /oak:index/nodetype*(1254)

[INFO] Reindexing completed
[INFO] Reindexing will be performed for following indexes: [/oak:index/principalName, /oak:index/authorizableId, /oak:index/acPrincipalName, /oak:index/repMembers]
[INFO] Indexing report
    - /oak:index/principalName*(2)
    - /oak:index/authorizableId*(2)
    - /oak:index/acPrincipalName*(0)
    - /oak:index/repMembers*(0)

[INFO] Reindexing completed
[INFO] Check report summary written to /Users/madamcin/oss/acs-aem-commons/content/target/oakpal-reports/oakpal-summary.json
[INFO] Evaluation of check reports has been deferred by 'deferBuildFailure=true'.
[INFO]
[INFO] --- oakpal-maven-plugin:1.1.3:verify (oakpal-verify) @ acs-aem-commons-content ---
[INFO] OakPAL Check Reports
[INFO]   net.adamcin.oakpal.core/basic/filterSets
[INFO]     +- <MINOR> non-default import mode MERGE defined for filterSet with root /var/acs-commons [acs-aem-commons-content-3.19.1-SNAPSHOT.zip]
[INFO]   com.adobe.acs.acs-aem-commons-oakpal-checks/aem64/content-classifications
[INFO]     +- <MINOR> /apps/acs-commons/components/authoring/graphiciconselect [protected super type]: /libs/granite/ui/components/foundation/form/select is marked FINAL [acs-aem-commons-content-3.19.1-SNAPSHOT.zip]
[INFO]     +- <MINOR> /apps/acs-commons/components/workflow/watson-audio-transcription [protected super type]: /libs/cq/workflow/components/model/external_process is marked FINAL [acs-aem-commons-content-3.19.1-SNAPSHOT.zip]
[INFO]     +- <MINOR> /apps/acs-commons/components/workflow/select-agent [protected super type]: /libs/cq/workflow/components/model/process is marked FINAL [acs-aem-commons-content-3.19.1-SNAPSHOT.zip]
[INFO]     +- <MINOR> /apps/acs-commons/touchui-widgets/contextualpathbrowser/.context [protected super type]: /libs/granite/ui/components/foundation/form/pathbrowser is marked FINAL [acs-aem-commons-content-3.19.1-SNAPSHOT.zip]
[INFO]     +- <MINOR> /etc/dam/video [protected resource type]: /libs/dam/components/video is marked INTERNAL [acs-aem-commons-content-3.19.1-SNAPSHOT.zip]
[INFO]     +- <MINOR> /etc/dam/video/mp3hq/jcr:content [protected resource type]: /libs/dam/components/video is marked INTERNAL [acs-aem-commons-content-3.19.1-SNAPSHOT.zip]
[INFO]     +- <MINOR> /etc/dam/video/ogghq/jcr:content [protected resource type]: /libs/dam/components/video is marked INTERNAL [acs-aem-commons-content-3.19.1-SNAPSHOT.zip]
[INFO]     +- <MINOR> /etc/dam/video/flacmono/jcr:content [protected resource type]: /libs/dam/components/video is marked INTERNAL [acs-aem-commons-content-3.19.1-SNAPSHOT.zip]
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for ACS AEM Commons Oakpal Checks 3.19.1-SNAPSHOT:
[INFO]
[INFO] ACS AEM Commons Oakpal Checks ...................... SUCCESS [  7.193 s]
[INFO] ACS AEM Commons Package ............................ SUCCESS [ 18.414 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  26.642 s
[INFO] Finished at: 2018-11-15T15:53:46-08:00
[INFO] ------------------------------------------------------------------------
```

### CODE REVIEW

**Comment by adamcin** [comment]
@joerghoh Here is that overlay check you were asking about.

**Comment by badvision** [comment]
I like it.  I'm just trying to grok what the best packaging is for this, given that it is a maven plugin ostensibly.  I'm trying to consider what it would look like to add this as a maven dependency by itself.  @davidjgonzalez, @justinedelson thoughts on this?

**Comment by justinedelson** [comment]
I feel like there are two different things being attempted here, although I may be misunderstanding.

1. A (very cool) Maven plugin is being added which does some validation. Usually a good thing.
2. A separate module is being created which is meant to be used by *consumers* of ACS AEM Commons to ensure that downstream packages don't conflict with ACS AEM Commons.

If that's the case, then I think it does make sense to put #2 as a module of the reactor, although I'm unclear why oakpal-checks/pom.xml is using the bundle packaging. Isn't this just used a build time and thus should be a regular JAR? Or does OakPAL use OSGi internally and thus the extensions must be bundles?

That said, I don't actually see anything related to ACS AEM Commons in oakpal-checks. These look like AEM rules. If there is desire for an OakPAL plugin (if that's the right term) to validate the AEM 6.4 rules, that's fine. That just doesn't belong (at least IMO) in ACS AEM Commons. Ideally, it would be produced by AEM proper and licensed appropriately. This honestly looks like a maintenance nightmare to me, especially given that the current rules have a number of as-of-yet-unresolved issues. What I would expect in an OakPAL plugin for ACS AEM Commons (i.e. for *consumers* of ACS AEM Commons) is something like "don't create service users in `/home/users/acs-commons`". That seems like something the ACS AEM Commons project can maintain quite easily.

Going back to the first item, I think adding validation rules are a great idea. But buggy ones are problematic. And unless this pull request is *also* going to fix those problems (or at least be able to mark them as false positives) and fail the build when problems are introduced in the further, I'm not sure what long term value it brings. It looks like something which can and will just get ignored at build time.

Put simply, as I commented in #1550, I don't think the content classifications in AEM 6.4 are correct (not unexpected for a v1) and wouldn't suggest that this (or any project) start making automatic decisions based on them.

**Comment by adamcin** [comment]
@justinedelson Yes, it is a bit of both. I would first clarify your second point a bit, in that it's not implementing checks for downstream package compliance with ACS AEM Commons, but is implementing sanity checks for the ACS AEM Commons Content package itself, with a side benefit of downstream projects potentially being able to use this checklist as well, maybe even to the level of a recommendation to use it. 

The question of whether the oakpal-checks module should be a bundle or just a jar is one I'm not particularly opinionated about. Oakpal checklists are namespaced using the library's module name, which currently can be derived from either a `Bundle-SymbolicName` or an `Automatic-Module-Name` header in the jar's manifest. I chose those headers out of convenience, but could certainly support an `Oakpal-ModuleName` header that doesn't imply that any OSGi or JPMS baggage is involved.

I think I understand your concerns about owning a package acceptance standard for AEM at large, and I think the best argument for that concern is the effort and potential licensing issues of hosting a full copy of the AEM product's nodetypes export. I don't know that I have a good answer for that, but maybe Adobe can fill that gap in the future by including all the product nodetypes in the uber-apis jars. For the question of maintaining that list of classified paths, I ended up just running a TraversingVisitor on /libs to construct the JSON array elements and pasted them in there.

Regarding the question of marking certain violations as false positives while also failing the build when new validations arise, this PR does implement configurable support for both of those requirements in the ContentClassifications check, by being able to exclude paths from the scope of the check, and by raising the severity to the minimum set for failing the build. The check implementation can be used multiple times in an execution with a different configuration each time.

I did just realize I missed the whole overlay aspect of the content classification rules, so I can iterate a bit more on this as it is, and try to address your other points too.

**Comment by adamcin** [comment]
@justinedelson I've made changes based on the some of the points you raised. I've renamed the checklist to indicate that its focus is limited to this project alone ("acs-commons-on-64"). The `oakpal-checks` module has been converted to `<packaging>jar</packaging>`. I've added exclusions in the checklist json file with comment attributes for the known conflicts with final and internal classifications. The build will fail any time a new classification conflict is added which is not covered by the existing excludes. Currently, the build is passing with the following output for `verify` on the `content` module:

```[INFO] --- oakpal-maven-plugin:1.1.4:scan (oakpal-scan-on-64) @ acs-aem-commons-content ---
[INFO] Found a new index node [reference]. Reindexing is requested
[INFO] Reindexing will be performed for following indexes: [/oak:index/uuid, /oak:index/reference, /oak:index/nodetype]
[INFO] Indexing report
    - /oak:index/uuid*(0)
    - /oak:index/reference*(0)
    - /oak:index/nodetype*(1254)

[INFO] Reindexing completed
[INFO] Reindexing will be performed for following indexes: [/oak:index/principalName, /oak:index/authorizableId, /oak:index/acPrincipalName, /oak:index/repMembers]
[INFO] Indexing report
    - /oak:index/principalName*(2)
    - /oak:index/authorizableId*(2)
    - /oak:index/acPrincipalName*(0)
    - /oak:index/repMembers*(0)

[INFO] Reindexing completed
[INFO] Check report summary written to /Users/madamcin/oss/acs-aem-commons/content/target/oakpal-reports/oakpal-summary.json
[INFO] Evaluation of check reports has been deferred by 'deferBuildFailure=true'.
[INFO]
[INFO] --- oakpal-maven-plugin:1.1.4:verify (oakpal-verify) @ acs-aem-commons-content ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for ACS AEM Commons Oakpal Checks 3.19.1-SNAPSHOT:
[INFO]
[INFO] ACS AEM Commons Oakpal Checks ...................... SUCCESS [  8.580 s]
[INFO] ACS AEM Commons Package ............................ SUCCESS [ 20.472 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  30.174 s
[INFO] Finished at: 2018-11-18T18:31:31-08:00
[INFO] ------------------------------------------------------------------------```

**Comment by justinedelson** [comment]
BTW, I think it would be pretty useful to have an OakPAL plugin for consumers of ACS AEM Commons. I'm not sure what entirely would be in those rules, but presumably there are things consumers should avoid. For example, consumers shouldn't create users whose names start with `acs-commons`.

**Comment by adamcin** [comment]
That makes sense to me. Maybe we could also warn on direct import of oak:index nodes and rep:Users in general, and use it as an opportunity to recommend using Ensure Oak Index and Ensure Authorizable, respectively.

**Comment by badvision** [comment]
This looks good to me, are we good to go with this one @justinedelson?

**Comment by adamcin** [comment]
@badvision @justinedelson Just added an integration test to check that the checklists themselves load.

**Comment by adamcin** [comment]
@justinedelson I added checks called RecommendEnsureAuthorizable and RecommendEnsureOakIndex as well, but I don't think I got your input on whether this kind of opinionated messaging would be an appropriate addition to this tool.

**Comment by adamcin** [comment]
@justinedelson @badvision Would it help if I were actively involved in the maintenance of this project?

**Comment by badvision** [comment]
Well, actively involved in bug/feature requests for this part at the very least.  I think it's a pretty compelling feature.  We work with a lot of geographically distributed teams and sometimes the only way to enforce standards is through automated checks like these.  If the code coverage is adequately asserting the behavior works as expected, and we can agree that any bugfixes related to this should be done via mutation testing (e.g. write test case to assert desired end state, then fix bug and ensure tests pass, etc.) then I'm happy to accept this PR. 😄

**Comment by badvision** [comment]
Chances are @HitmanInWis will come after you before I do anyway if there's bugs in it.  So 👍 from me. 😉

**Comment by kwin** [comment]
I still have issues to understand how the content classification check (https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/master/oakpal-checks/src/main/java/com/adobe/acs/commons/oakpal/checks/ContentClassifications.java) works internally. Against which repo does it check? How do I configure the baseline? What is used as current baseline in acs-aem-commons?

**Comment by justinedelson** [comment]
@kwin I'd recommend reading both the OakPAL documentation and watching the AEM Rockstar session from Adobe Summit (https://summit.adobe.com/na/summit-online/#26579). That explains the architecture of OakPAL reasonably well.

The short answer is that OakPAL creates a new Oak repository and then installs packages into it (in this case the ACS AEM Commons package generated by the build). While installing, a listener is firing off events as nodes are created/modified/deleted and then OakPAL rules are evaluating those events.

I'm not sure what you mean by "baseline".

**Comment by kwin** [comment]
@justinedelson The baseline is the existing nodes in the repository in which the package is installed. I updated above that this is maintained in https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/88a9bec43adc3f39516e500a55e3d46e315739b6/oakpal-checks/src/main/resources/OAKPAL-INF/checklist/acs-internal.json#L62. But I am really wondering how to update this? This is the metadata of roughly 1000 nodes. There should be some hints somewhere on how to retrieve that automatically from a running AEM instance....

**Comment by justinedelson** [comment]
ah, so your question is how the checklist gets generated? Created #1872 for that
